rfcs/0642-app-invocation-parameters.md

```diff
@@ -0,0 +1,92 @@
+                             ^
+                             |
+Alert #2 [count < 5] --------+
+```
+
+This effectively prevents non-administrative users from making full use of alerts, since "alert-specific" app configuration, like a recipient email address, is bundled up with inaccessible infrastructure details like SMTP server hostnames and passwords.
+
+By marking specific app settings as user-configurable, this feature will make it possible for (unprivileged) users to set up their own alerts, and reduce the amount of duplicated app instance configuration.
+
+```
+Alert #1 [count > 1, user1@example.com] ---> Email app instance [_, smtp.example.com, pa$$word]
+                                                ^
+                                                |
+Alert #2 [count < 5, user2@example.com] --------+
+```
+
+## Detailed design
+
+As implemented today, apps expose _available settings_, which each carry a unique name (like `"EmailAddress"`), an input type (e.g. `Text`) and other details. When an app instance is created, it is given a dictionary of _settings values_, keyed by setting name. For example, `("EmailAddress", "user@example.com")`.
+
+The design proposes that app instances carry a new list of _invocation parameters_. For each app setting, the setting name will **either** be given a setting value, **or** appear in the instance's invocation parameters. An app instance that has one or more invocation parameters is said to be _parameterized_.
+
+From a user's perspective, parameterized app instances are just normal app instances, but they behave slightly differently from existing apps.
+
+1. When a notification is configured to use a parameterized app instance, values must be provided for the instance's invocation parameters
+1. Likewise, when sending an event manually to a parameterized app instance, invocation parameter values must be supplied
+1. A parameterized app instance cannot receive events directly from the stream
+
+### App instance configuration
+
+There are two requirements imposed by this feature:
+
+1. Parameterized app instances can't be configured to receive events directly from the stream
+1. Individual settings need to be marked as invocation parameters
+
+Rather than use the presence of invocation parameters (2.) to drive the application of rule (1.), this design proposes that we allow parameterization to occur only when the app is marked as "manual input only". In 4.2, this setting is under the _Input_ heading, and called _Stream incoming events_.
+
+This adds additional semantics to the existing field, but these are compatible with the current behavior of this setting.
+
+The setting label and help text will be:
+
+> **Stream incoming events**
+> App instances can be triggered by alerts set on dashboards, and events can be sent manually to any app from the drop-down menu in the event details. Check this box to send events as they arrive.
```

Comment:
I think the sentence 
> Check this box to send events as they arrive

is a little confusing, mostly the `send events` which, for an app like *EmailPlus* sounds like we are sending the events as e-mail. 

Also I think the name `Stream incoming events` is a little vague.

I can't really find better wording right now, though :p